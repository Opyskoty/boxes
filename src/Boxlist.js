import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import uuid from "uuid/v4";

function Boxlist() {

  const [list, setList] = useState([]);

  const addItem = item => {
    let newItem = { ...item, id: uuid() }
    setList(list => [...list, newItem]);
  }

  const removeItem = id => {
    setList(list => list.filter(b => b.id !== id));
  }

  return (
    <div className="Boxlist">
      <NewBoxForm addItem={addItem} />
      <ul>
        {list.map(b => <Box {...b} removeItem={removeItem} />)}
      </ul>
    </div>
  )
}

export default Boxlist;