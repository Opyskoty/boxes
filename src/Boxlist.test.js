import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Boxlist from "./Boxlist";

it("renders without crashing", () => {
  render(<Boxlist />)
});

it("matches the snapshot", () => {
  const { asFragment } = render(<Boxlist />);

  expect(asFragment()).toMatchSnapshot();
});

it("has the ability to enter data", () => {
  const { getByLabelText, queryByText, queryByTestId } = render(<Boxlist />);

  // no box yet
  expect(queryByTestId("box")).not.toBeInTheDocument();

  const widthInput = getByLabelText("Width");
  const heightInput = getByLabelText("Height");
  const backgroundColorInput = getByLabelText("Background Color");
  const submitButton = queryByText("Add a Box!");

  fireEvent.change(widthInput, { target: { value: 100 } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(backgroundColorInput, { target: { value: "yellow" } });
  fireEvent.click(submitButton);

  expect(queryByTestId("box")).toBeInTheDocument();
});

it("has the ability to remove a box", () => {
  const { getByLabelText, queryByText, queryByTestId } = render(<Boxlist />);

  // no box yet
  expect(queryByTestId("box")).not.toBeInTheDocument();

  const widthInput = getByLabelText("Width");
  const heightInput = getByLabelText("Height");
  const backgroundColorInput = getByLabelText("Background Color");
  const submitButton = queryByText("Add a Box!");

  fireEvent.change(widthInput, { target: { value: 100 } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(backgroundColorInput, { target: { value: "yellow" } });
  fireEvent.click(submitButton);

  expect(queryByTestId("box")).toBeInTheDocument();

  const removeButton = queryByText("X");
  fireEvent.click(removeButton);

  expect(queryByTestId("box")).not.toBeInTheDocument();
});


it("has the ability add multiple boxes", () => {
  const { getByLabelText, queryByText, queryByTestId, queryAllByTestId } = render(<Boxlist />);

  // no box yet
  expect(queryByTestId("box")).not.toBeInTheDocument();

  const widthInput = getByLabelText("Width");
  const heightInput = getByLabelText("Height");
  const backgroundColorInput = getByLabelText("Background Color");
  const submitButton = queryByText("Add a Box!");

  fireEvent.change(widthInput, { target: { value: 100 } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(backgroundColorInput, { target: { value: "yellow" } });
  fireEvent.click(submitButton);

  fireEvent.change(widthInput, { target: { value: 300 } });
  fireEvent.change(heightInput, { target: { value: 300 } });
  fireEvent.change(backgroundColorInput, { target: { value: "pink" } });
  fireEvent.click(submitButton);

  fireEvent.change(widthInput, { target: { value: 400 } });
  fireEvent.change(heightInput, { target: { value: 400 } });
  fireEvent.change(backgroundColorInput, { target: { value: "blue" } });
  fireEvent.click(submitButton);

  expect(queryAllByTestId("box").length).toBe(3);
});