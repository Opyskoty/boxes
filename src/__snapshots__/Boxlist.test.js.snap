// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Boxlist"
  >
    <form>
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value="0"
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value="0"
      />
      <label
        for="backgroundColor"
      >
        Background Color
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add a Box!
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
